replay/models/nmslib_hnsw.py

```diff
@@ -0,0 +1,646 @@
+                method=self._method,
+                space=self._space,
+                data_type=nmslib.DataType.SPARSE_VECTOR,  # pylint: disable=c-extension-no-member
+            )
+            if isinstance(self._index_file, FileInfo):
+                load_index_from_source_fs(
+                    sparse=True,
+                    load_index=lambda path: self._index.loadIndex(
+                        path, load_data=True
+                    ),
+                    source=self._index_file
+                )
+            else:
+                self._index.loadIndex(
+                    SparkFiles.get(self._index_file), load_data=True
+                )
+        else:
+            self._index = nmslib.init(  # pylint: disable=c-extension-no-member
+                method=self._method,
+                space=self._space,
+                data_type=nmslib.DataType.DENSE_VECTOR,  # pylint: disable=c-extension-no-member
+            )
+            if isinstance(self._index_file, FileInfo):
+                load_index_from_source_fs(
+                    sparse=False,
+                    load_index=lambda path: self._index.loadIndex(path),  # pylint: disable=unnecessary-lambda
+                    source=self._index_file
+                )
+            else:
+                self._index.loadIndex(SparkFiles.get(self._index_file))
+
+        if self._ef_s:
+            self._index.setQueryTimeParams({"efSearch": self._ef_s})
+        return self._index
+
+
+class NmslibHnswMixin(ANNMixin):
+    """Mixin that provides methods to build nmslib hnsw index and infer it.
+    Also provides methods to saving and loading index to/from disk.
+    """
+
+    def _infer_ann_index(  # pylint: disable=too-many-arguments
+        self,
+        vectors: DataFrame,
+        features_col: str,
+        params: Dict[str, Union[int, str]],
+        k: int,
+        filter_seen_items: bool,
+        index_dim: str = None,
+        index_type: str = None,
+        log: DataFrame = None,
+    ) -> DataFrame:
+        return self._infer_nmslib_hnsw_index(
+            vectors,
+            features_col,
+            params,
+            k,
+            filter_seen_items,
+            index_type,
+            log,
+        )
+
+    def _build_ann_index(  # pylint: disable=too-many-arguments
+        self,
+        vectors: DataFrame,
+        features_col: str,
+        params: Dict[str, Union[int, str]],
+        dim: int = None,
+        num_elements: int = None,
+        id_col: Optional[str] = None,
+        index_type: str = None,
+        items_count: Optional[int] = None,
+    ) -> None:
+        self._build_nmslib_hnsw_index(
+            vectors, features_col, params, index_type, items_count
+        )
+
+    def _build_nmslib_hnsw_index(  # pylint: disable=too-many-arguments, too-many-locals, too-many-statements
+        self,
+        item_vectors: DataFrame,
+        features_col: str,
+        params: Dict[str, Any],
```

Comment:
why `index_type, items_count` are not included into `params`

Reply:
`index_type` removed. Maybe `items_count` should be moved into the new `NmslibHnswParam` class, I don't know yet.